dnsclient/dnsclient.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dnsclient is a client library for performing CT operations
+// over DNS.  The DNS mechanism is experimental and subject to change.
+package dnsclient
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base32"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/tls"
+)
+
+// DNSClient represents a DNS client for a given CT Log instance
+type DNSClient struct {
+	base     string
+	Verifier *ct.SignatureVerifier // nil for no verification (e.g. no public key available)
+	resolve  func(ctx context.Context, name string) ([]string, error)
+}
+
+// New constructs a new DNSClient instance.  The base parameter gives the
+// top-level domain name; opts can be used to provide a custom logger
+// interface and a public key for signature verification.
+func New(base string, opts jsonclient.Options) (*DNSClient, error) {
+	pubkey, err := opts.ParsePublicKey()
+	if err != nil {
+		return nil, fmt.Errorf("invalid public key: %v", err)
+	}
+
+	var verifier *ct.SignatureVerifier
+	if pubkey != nil {
+		var err error
+		verifier, err = ct.NewSignatureVerifier(pubkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(base) > 0 && base[len(base)-1] != '.' {
+		base += "."
+	}
+	return &DNSClient{
+		base:     base,
+		Verifier: verifier,
+		resolve:  func(ctx context.Context, name string) ([]string, error) { return net.LookupTXT(name) },
+	}, nil
+}
+
+const (
+	// Match base64 data (although note that this will also match invalid data with
+	// padding in the middle rather than the end).
+	base64RE = "[A-Za-z0-9+/=]+"
```

Comment:
Wouldn't be too difficult to change the regex to accept the '=' character only at the end? Just a suggestion, feel free to ignore. :stuck_out_tongue: